### ISSUE TITLE
muntjac-soc firmware build issue

### ISSUE THREAD

**Issue by SazabiUC** [nbdd0121/muntjac-soc#2]
Hi, I'm getting firmware build issue from compiling unwinding during building the firmware for muntjac-soc. I tried to set different release version and commit version of unwinding, still build fails. 

error[E0433]: failed to resolve: could not find `video` in the crate root
  --> src/fmt.rs:26:46
   |
26 |         if let Some(fbcon) = unsafe { crate::video::get_fbcon() } {
   |                                              ^^^^^ could not find `video` in the crate root

**Comment by nbdd0121** [comment]
I forget to cfg it with `has_display`. A fix has been pushed. Can you retry?

**Comment by SazabiUC** [comment]
Hi, thanks, I just retried, the error seems to disappear. But another error appeared regarding linking the bootloader. I'm building the genesis 2 board by the way. 

error: linking with `/home/yue/repo/muntjac/muntjac-soc/firmware/../build/linker` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/yue/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/yue/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/yue/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/tools/modelsim/modelsim_dlx/bin:/tools/verilator/v4.210/bin:/home/yue/intelFPGA/20.1/modelsim_ase/bin:/home/yue/Documents/riscv/lowrisc-toolchain-gcc-rv32imcb-20240206-1/bin:/home/yue/Documents/bluespec/bsc-2023.07-ubuntu-20.04/bin/:/tools/Xilinx/Vitis/2022.2/bin/:/tools/Xilinx/Vivado/2022.2/bin/:/home/yue/.cargo/bin:/home/yue/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin" VSLANG="1033" "/home/yue/repo/muntjac/muntjac-soc/firmware/../build/linker" "/tmp/rustcw3assH/symbols.o" "/home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/bootloader-9e052eecca060d1f.bootloader.e2946d6e6a55e1b1-cgu.5.rcgu.o" "--as-needed" "-L" "/home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps" "-L" "/home/yue/repo/muntjac/muntjac-soc/genesys_2/build/release/deps" "-L" "/home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/build/bootloader-be22c16e02a4970d/out" "-L" "/home/yue/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/riscv64imac-unknown-none-elf/lib" "-Bstatic" "-lfoo" "/home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib" "-Bdynamic" "-z" "noexecstack" "-L" "/home/yue/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/riscv64imac-unknown-none-elf/lib" "-o" "/home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/bootloader-9e052eecca060d1f" "--gc-sections" "-O1" "-Tlinker.ld" "--eh-frame-hdr"
  = note: riscv64-linux-gnu-ld: -march=rv64i2p1_m2p0_a2p1_f2p2_d2p2_c2p0_zicsr2p0_zifencei2p0: unsupported ISA subset `z'
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.06.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.10.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.11.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.12.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.13.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.14.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.15.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.00.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.01.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.02.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.03.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.04.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.05.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.07.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.08.rcgu.o)
          riscv64-linux-gnu-ld: -march=: ISA string must begin with rv32 or rv64
          riscv64-linux-gnu-ld: failed to merge target specific data of file /home/yue/repo/muntjac/muntjac-soc/genesys_2/build/riscv64imac-unknown-none-elf/release/deps/libcompiler_builtins-f8635da2e44ebe0b.rlib(compiler_builtins-f8635da2e44ebe0b.compiler_builtins.804125e885f6555f-cgu.09.rcgu.o)
          

warning: `bootloader` (bin "bootloader") generated 4 warnings
The following warnings were emitted during compilation:

warning: bootloader@0.1.0: src/entry.S: Assembler messages:
warning: bootloader@0.1.0: src/entry.S:4: Warning: Unrecognised .option directive: arch,+zicsr
warning: bootloader@0.1.0: 

error: could not compile `bootloader` (bin "bootloader") due to 1 previous error; 4 warnings emitted

**Comment by nbdd0121** [comment]
What's the version of the RISC-V GCC toolchain are you using?  `riscv64-linux-gnu-gcc --version` should report it. I think it's very likely that you're using a very old toolchain.

**Comment by SazabiUC** [comment]
Yes, you were right, the gcc installed from sudo apt-get install gcc-riscv64-linux-gnu was quite an old version, was 2019. I had to build it from source, using that one successfully compiled. Thanks for your help and your great works on Muntjac. I really like it.

**Comment by SazabiUC** [comment]
Hi, I got another issue which is about building rootfs.img, I've tried to build it on multiple Ubuntu machines all failed, getting the same error. Haven't got fixed yet, would be appreciate if you could please look into it.

I: content of /etc/apt/sources.list:
deb http://deb.debian.org/debian-ports/ unstable main
E: apt-get update didn't download anything
I: Updating apt sources
chroot: failed to run command ‘apt-get’: No such file or directory
I: Install additional packages
chroot: failed to run command ‘apt-get’: No such file or directory
I: Patching
sed: can't read /tmp/tmp.hYKtvdPVjJ/usr/bin/which.debianutils: No such file or directory
I: Unmounting rootfs image

**Comment by nbdd0121** [comment]
Debian has recently moved riscv64 from port to their officially supported arch. I have updated the URL in the latest commit. Can you try again?

**Comment by SazabiUC** [comment]
Hi, thanks for looking into it. I just pulled and tried it again. But got a different error. I looked it up on the web, tried all the solutions they suggested, seemed not fixed the issue.

Initializing machine ID from random generator.
Failed to take /etc/passwd lock: Invalid argument
dpkg: error processing package systemd (--configure):
 installed systemd package post-installation script subprocess returned error exit status 1
dpkg: dependency problems prevent configuration of cron-daemon-common:
 cron-daemon-common depends on systemd | systemd-standalone-sysusers | systemd-sysusers; however:
  Package systemd is not configured yet.
  Package systemd-standalone-sysusers is not installed.
  Package systemd-sysusers is not installed.
  Package systemd which provides systemd-sysusers is not configured yet.

dpkg: error processing package cron-daemon-common (--configure):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 systemd
 cron-daemon-common
E: Sub-process /usr/bin/dpkg returned an error code (1)
E: run_chroot failed: E: /usr/sbin/chroot /tmp/tmp.2AXAoZP5Uh env --unset=APT_CONFIG --unset=TMPDIR apt-get --yes install -oAPT::Status-Fd=<$fd> -oDpkg::Use-Pty=false libpam-modules systemd debian-archive-keyring iproute2 readline-common iputils-ping vim-tiny mawk apt-utils passwd cron-daemon-common e2fsprogs login sensible-utils kmod fdisk dmidecode tzdata ifupdown adduser nftables less mount nano systemd-sysv gpgv logrotate netbase udev procps init apt cpio vim-common tasksel-data libpam-modules-bin isc-dhcp-common whiptail debconf-i18n cron isc-dhcp-client libpam-runtime debconf failed
